File: program/src/integrations/drift/pdas.rs
use pinocchio::{
    program_error::ProgramError,
    pubkey::{try_find_program_address, Pubkey},
};

use crate::integrations::drift::constants::DRIFT_PROGRAM_ID;

/// Derive the Drift User PDA from subaccount
pub fn derive_drift_user_pda(
    controller_authority: &Pubkey,
    sub_account_id: u16,
) -> Result<Pubkey, ProgramError> {
    let (pda, _) = try_find_program_address(
        &[
            b"user",
            controller_authority.as_ref(),
            &sub_account_id.to_le_bytes(),
        ],
        &DRIFT_PROGRAM_ID,
    )
    .ok_or(ProgramError::InvalidSeeds)?;
    Ok(pda)
}

/// Derive the Drift SpotMarket PDA
pub fn derive_drift_spot_market_pda(market_index: u16) -> Result<Pubkey, ProgramError> {
    let (pda, _) = try_find_program_address(
        &[b"spot_market", market_index.to_le_bytes().as_ref()],
        &DRIFT_PROGRAM_ID,
    )
    .ok_or(ProgramError::InvalidSeeds)?;
    Ok(pda)
}

/// Derive SpotMarket Vault PDA
pub fn derive_drift_spot_market_vault_pda(market_index: u16) -> Result<Pubkey, ProgramError> {
    let (pda, _) = try_find_program_address(
        &[b"spot_market_vault", market_index.to_le_bytes().as_ref()],
        &DRIFT_PROGRAM_ID,
    )
    .ok_or(ProgramError::InvalidSeeds)?;

    Ok(pda)
}

/// Derive Drift UserStats PDA
pub fn derive_drift_user_stats_pda(authority: &Pubkey) -> Result<Pubkey, ProgramError> {
    let (pda, _) =
        try_find_program_address(&[b"user_stats", authority.as_ref()], &DRIFT_PROGRAM_ID)
            .ok_or(ProgramError::InvalidSeeds)?;

    Ok(pda)
}

/// Derives Drift State PDA
pub fn derive_drift_state_pda() -> Result<Pubkey, ProgramError> {
    let (pda, _) = try_find_program_address(&[b"drift_state"], &DRIFT_PROGRAM_ID)
        .ok_or(ProgramError::InvalidSeeds)?;

    Ok(pda)
}

File: clients/ts/src/integrations/drift/pdas.ts
import {
  address,
  Address,
  getAddressEncoder,
  getU16Encoder,
  getProgramDerivedAddress,
} from "@solana/addresses";

export const DRIFT_PROGRAM_ID =
  "dRiftyHA39MWEi3m9aunc5MzRF1JYuBsbn6VPcn33UH" as Address;

/// Derives Drift CPI signer
export const deriveDriftSigner = async () => {
  const [driftSignerPda] = await getProgramDerivedAddress({
    programAddress: DRIFT_PROGRAM_ID,
    seeds: ["drift_signer"],
  });

  return driftSignerPda;
};

/// Derives Drift CPI signer nonce
export const deriveDriftSignerNonce = async (): Promise<number> => {
  const [, nonce] = await getProgramDerivedAddress({
    programAddress: DRIFT_PROGRAM_ID,
    seeds: ["drift_signer"],
  });

  return nonce;
};

/// Derives State PDA
export const deriveStatePda = async () => {
  const [statePda] = await getProgramDerivedAddress({
    programAddress: DRIFT_PROGRAM_ID,
    seeds: ["drift_state"],
  });

  return statePda;
};

/// Derives UserStats PDA
export const deriveUserStatsPda = async (authority: Address) => {
  const addressEncoder = getAddressEncoder();

  const [userStatsPda] = await getProgramDerivedAddress({
    programAddress: DRIFT_PROGRAM_ID,
    seeds: ["user_stats", addressEncoder.encode(authority)],
  });

  return userStatsPda;
};

/// Derives User subaccount PDA
export const deriveUserPda = async (
  authority: Address,
  subAccountId: number
) => {
  const addressEncoder = getAddressEncoder();

  const [userPda] = await getProgramDerivedAddress({
    programAddress: DRIFT_PROGRAM_ID,
    seeds: [
      "user",
      addressEncoder.encode(authority),
      getU16Encoder
      .encode(subAccountId),
    ],
  });

  return userPda;
};

/// Derives SpotMarket PDA
export const deriveSpotMarketPda = async (marketIndex: number) => {
  const [spotMarketPda] = await getProgramDerivedAddress({
    programAddress: DRIFT_PROGRAM_ID,
    seeds: [
      "spot_market",
      getU16Encoder.encode(marketIndex),
    ],
  });

  return spotMarketPda;
};

/// Derives SpotMarket Vault PDA
export const deriveSpotMarketVaultPda = async (marketIndex: number) => {
  const [spotMarketVaultPda] = await getProgramDerivedAddress({
    programAddress: DRIFT_PROGRAM_ID,
    seeds: [
      "spot_market_vault",
      getU16Encoder.encode(marketIndex),
    ],
  });

  return spotMarketVaultPda;
};
